src/fides/api/util/tcf_util.py

```diff
@@ -0,0 +1,637 @@
+    information underneath.  Likewise, puts vendor and system information top-level, and embeds purpose and feature information
+    underneath.
+
+    System map is passed in as an argument instead of being constructed here, because both
+    purposes/features are added to it and we modify that mapping multiple times.
+    """
+    matching_purpose_or_feature_map: Dict[int, TCFPurposeOrFeature] = {}
+
+    is_purpose_section: bool = "purposes" in tcf_component_name
+
+    for privacy_declaration_row in matching_privacy_declaration_query:
+        # Filter relevant uses or features, depending on the section of the tcf overlay
+        relevant_use_or_features: Set[str] = get_matching_data_uses_or_features(
+            record=privacy_declaration_row,
+            relevant_uses_or_features=relevant_uses_or_features,
+            is_purpose_type=is_purpose_section,
+        )
+
+        for attribute in relevant_use_or_features:
+            # Add top-level entry to purpose or feature section if applicable
+            top_level_tcf_record: Optional[
+                TCFPurposeOrFeature
+            ] = _add_top_level_record_to_purpose_or_feature_section(
+                matching_purpose_or_feature_map=matching_purpose_or_feature_map,
+                is_purpose_type=is_purpose_section,
+                use_or_feature=attribute,
+            )
+
+            if not top_level_tcf_record:
+                continue
+
+            vendor_id, system_identifier = get_system_identifiers(
+                privacy_declaration_row
+            )
+            # Add top-level entry to the system section if applicable
+            if system_identifier not in system_map:
+                system_map[system_identifier] = TCFVendorRecord(
+                    id=system_identifier,  # Identify system by vendor id if it exists, otherwise use system id.
+                    name=privacy_declaration_row.system_name,
+                    description=privacy_declaration_row.system_description,
+                    has_vendor_id=bool(
+                        vendor_id
+                    ),  # Has_vendor_id will let us separate data into two sections: "tcf_vendors" and "tcf_systems"
+                )
+
+            # Add the purpose or feature to the System purposes or System features lists
+            embedded_purpose_or_feature_record: TCFPurposeOrFeature = (
+                _clone_top_level_record_then_add_legal_bases(
+                    top_level_tcf_record,
+                    privacy_declaration_row.legal_basis_for_processing,
+                )
+            )
+
+            _embed_purpose_or_feature_under_system(
+                embedded_tcf_record=embedded_purpose_or_feature_record,
+                system_section=getattr(
+                    system_map[system_identifier], tcf_component_name
+                ),
+                legal_basis_for_processing=privacy_declaration_row.legal_basis_for_processing,
+            )
```

Comment:
nit: if i'm following correctly, this is a bit redundant - the `privacy_declaration_row.legal_basis_for_processing` has already been added to the `embedded_purpose_or_feature_record` object when it was instantiated within ` _clone_top_level_record_then_add_legal_bases`. 

when reading over ` _clone_top_level_record_then_add_legal_bases`, it struck me as a bit of a "random" assortment of operations. i think you could probably simplify things by removing that function and slightly refactoring things here:

```suggestion
            # copy top level record to preserve its initial state to use for embedded record
            embedded_tcf_record: TCFPurposeOrFeature = top_level_tcf_record.copy()
            # Override legal_bases on the embedded purpose or special purpose record
            _extend_legal_bases(embedded_tcf_record, legal_basis_for_processing, replace=True)
            
            _embed_purpose_or_feature_under_system(
                embedded_tcf_record=embedded_purpose_or_feature_record,
                system_section=getattr(
                    system_map[system_identifier], tcf_component_name
                ),
            )
            
            # Append to the existing legal_bases on the top-level record if applicable for purpose and special purpose sections.
            # The top-level record is updated in-place
            _extend_legal_bases(top_level_tcf_record, legal_basis_for_processing, replace=False)
```

and then within `_embed_purpose_or_feature_under_system` just decide whether to add the `legal_basis_for_processing` that's on the provided embedded purpose to the _existing_ embedded purpose (if it exists), or to add it to the "new" embedded purpose (i.e. the provided one) to the system -- basically the same as it does currently.

does that makes sense? if not, happy to provide a commit that makes the change, maybe that'd be simpler. if it makes sense and you think i'm mistaken, then sorry for the rant 😄 but i do think this may deserve a slight refactor to make it easier to follow, in any case!

Reply:
I agree with you this could be more readable for sure.  I hard time layering in the legal basis for processing bit in a clean way:

The main motivation for passing in the legal basis here when I embed the purpose under a vendor, is that the purpose might already exist under that vendor.   If the purpose is already there, then I would consolidate the legal_bases across both purposes. It's a bit contrived but seems possible, especially given that users can set up their own systems, it is not just pristine GVL systems that appear.

Reply:
right i think i'm following that, i don't think it's wrong that you're account for that contrived use case. i realize now though i messed up the wording in my comment a bit and i wasn't very clear 🤦 

i'm getting into the weeds and it's probably not worth much more of our time, since this is a stylistic concern, not functional. your most recent changes are close to what i was suggesting, but could we remove the `legal_basis_for_processing` arg from the ` _embed_purpose_or_feature_under_system` function, and instead within that function just read the legal basis from the provided `embedded_tcf_record`? or alternatively we could remove the 
```
 _extend_legal_bases(
                embedded_tcf_record, legal_basis_for_processing, replace=True
            )
```
that's done before the `_embed_purpose_or_feature_under_system` call and handle extending the legal bases _within_ the `_embed_purpose_or_feature_under_system` function. 

i'll make a code suggestion on your most recent revision, if it doesn't make sense then i think we can just leave it. i find the current code a _bit_ confusing but i don't think this will have any functional impact and it's a nit in any case.

Reply:
oh sure I think I follow what you mean now